### DISCUSSION TITLE
各paramについてのメモ

### DISCUSSION THREAD

**Discussion by Naoki-Hiraoka** [YutaKojio/hrpsys-base#69] [Q&A]
スライドにも資料があって重複していますが，僕の手元のメモを共有します．
このリストにないparamは把握しなくてよいかなと考えているのですが，このリストにないparamで，押さえておくべきparamがあれば教えて下さい

# AutoBalancer param
- 一般
  - transition_time : [s] for start and stop AutoBalancer
  - leg_names : 両脚か片脚か. 両脚なら歩行可
- 重心軌道関連
  - default_zmp_offsets : 各EEFの目標ZMPのオフセット. 滑らかに補間される
  - zmp_transition_time : [s] for default_zmp_offsets
- IK関連
  - use_collision_avoidance : whether use collision avoidance
  - cog_z_constraint : cog height constraint in fullbody IK
  - ik_max_loop : ikのループ数
  - is_natural_walk : 足の動きに合わせて肩を動かす
  - arm_swing_deg : is_natural_walkで肩を動かす大きさ
- 遊脚軌道関連
  - is_stop_early_foot : 早付き対応するか
- hand関連
  - use_force_mode : armのref_forceに応じて重心を動かすか
  - is_hand_fix_mode : Flag for fix hand while walking.
  - graspless_manip_mode
  - graspless_manip_arm
  - graspless_manip_p_gain
  - graspless_manip_reference_trans_pos
  - graspless_manip_reference_trans_rot
- emergency関連
  - is_emergency_step_mode : whether step when emergency
  - is_emergency_touch_wall_mode : whether touch wall when emergency
  - touch_wall_retrieve_time : Sync time for retrieve touch wall motion [s]
- toe Heel 関連
  - toe-angle
  - heel-angle
  - toe-heel-phase-ratio
  - use-toe-heel-transition
  - use-toe-heel-auto-set

# GaitGenerator Param
- 一般
  - gravitational_acceleration : Gravitational acceleration [m/s^2]
- 支持領域関連
  - leg_margin : Leg margin between foot end effector position and foot edge (front, rear, outside, inside) [m]. stride_limitation時の干渉判定と、CP転倒判定の閾値として用いる
  - safe_leg_margin : Safe leg margin between foot end effector position and foot edge (front, rear, outside, inside) [m]. その他.
- footstep関連
  - default_step_time : step time[s]. goVelocity, goPosの初期目標1歩の時間.
  - default_double_support_ratio : Ratio of double support period. Ratio is included in (0, 1). Double support period time is default_double_support_ratio*default_step_time. この値を2分割してstepの前後に両足支持期を設定する
  - stride_limitation_type: CIRCLEタイプ以外のモードはあまり使われない．
  - stride_limitation_for_circle_type : (forward, outside, theta, backward, inside) [m]. (反対の脚EEFからの距離. insideは遊脚のvertexが支持脚EEFからこの値以下に入らない). goVelocity, goPosの初期目標着地位置に用いる.
  - is_slow_stair_mode : srcとdstの高さがdefault_step_height - 0.02以上変化するなら， stair_step_timeだけ一歩の時間をdefault_step_timeよりも長くする
  - stair_step_time : 同上
- 重心軌道関連
  - dcm_offset : Offset of y component of ref_dcm. (CapturePointを内側にどれだけよせるか。終端CP位置や、着地位置修正判定などに用いる
  - use_act_states : ZMP計算でact stateを使うかどうか
  - is_interpolate_zmp_in_double : 
  - num_preview_step : horizonの大きさ
- 遊脚軌道関連
  - default_step_height : Step height [m]. startとgoalの高い方よりもこの値だけ上に
  - swing_trajectory_delay_time_offset : 遊脚軌道中の、time_offset後の位置にtime_offset後に到達するようにhofferbib平滑化する
  - default_orbit_type: 遊脚軌道としてrectangle以外のモードはあまり使われていない
  - rectangle_trajectory_way_point_offset : 中間の点の位置のoffset. 
  - rectangle_goal_off : goalの地面へのめり込み量
  - rectangle_time_smooth_offset : to make xy swing movement slower at first 
- 着地位置修正関連
  - modify_footsteps : Whether modify footsteps based on Capture Point
  - overwritable_footstep_index_offset : Offset for overwritable footstep index. Offset from current footstep index. Used in emergency_stop and velocity_mode.
  - overwritable_stride_limitation : Stride limitation when overwriting footsteps (forward, outside, theta, backward, inside) [m]
  - margin_time_ratio : Margin time ratio for footstep modification before landing. 着地前のこの間(0~1)は着地時間・位置修正を行わない
  - min_time_mgn : overwritable min time until landing for fg[s]. 次indexまでの残り時間がこの値を下回るようには着地時間修正を行わない
  - min_time : overwritable min step time for fg[s]. 一歩の時間(step_time)がこの値を下回るようには着地時間修正を行わない
  - overwritable_max_time : 一歩の時間(step_time)がこの値を上回るようには着地時間修正を行わない
  - fg_cp_cutoff_freq :  現在の実機のCapture Pointへのローパスフィルタ
- 角運動量
  - use_flywheel_balance : 角運動量を利用するかどうか
- 長期的外乱補償
  - use_disturbance_compensation : Whether use disturbance compensation or not
  - dc_gain : Gain for disturbance compensation
- goPos関連
  - optional_go_pos_finalize_footstep_num : Number of optional finalize footsteps in goPos
  - is_stable_go_stop_mode : whether gostop after becoming stable
- emergency step関連
  - emergency_step_time : Step time when emergency stepping [s]
- stuck判定関連
  - sum_d_footstep_thre
  - footstep_check_delta
# Stabilizer param
- 一般
  - eefm_gravitational_acceleration : 重力加速度
  - transition_time: モード遷移の時間
  - use_force_sensor : trueだと、on_ground判定を行う. falseだと、leg以外はact_contact_stateとしてref_contact_stateを使う
- filter関連
  - eefm_zmp_delay_time_const : Time constant for stabilizer ZMP delay [s] (x,y)
  - eefm_cogvel_cutoff_freq : Cutoff frequency of LPF in calculation of COG velocity [Hz]
- ZMP Distribution関連
  - eefm_wrench_alpha_blending : Blending parameter [0, 1] for wrench distribution
  - eefm_alpha_cutoff_freq : cutoff frequency of LPF in calculation of force moment distribution alpha ratio parameter [Hz]
  - eefm_ee_forcemoment_distribution_weight : Sequence of all end-effector force/moment distribution weight
  - st_algorithm : Current Stabilizer algorithm
- Damping Control関連
  - is_feedback_control_enable : Bool sequence for all end effectors whether the end effector is used for feedback control (currently damping control).
  - eefm_rot_damping_gain : Sequence of all end-effector rotation damping gain [Nm/(rad/s)] (r,p,y).
  - eefm_swing_rot_damping_gain : Rotation damping gain for swing leg [Nm/(rad/s)] (r,p,y)
  - eefm_rot_time_const : Sequence of all end-effector rotation damping time constant [s] (r,p,y)
  - eefm_pos_damping_gain : Sequence of all end-effector position damping gain [N/(m/s)] (x,y,z).
  - eefm_swing_pos_damping_gain : Position damping gain for swing leg [N/(m/s)] (x,y,z)
  - eefm_pos_time_const_support : Sequence of all end-effector position damping time constant for double support phase [s] (x,y,z).
  - eefm_pos_time_const_swing : End-effector position damping time constant for single support phase [s].
  - eefm_pos_transition_time : Transition time for single=>double support phase gain interpolation [s].
  - eefm_ee_moment_limit : Sequence of all end-effector end-effector-frame moment limit [Nm]
  - eefm_pos_compensation_limit : Sequence of all end-effector position compensation limit [m]
  - eefm_rot_compensation_limit : Sequence of all end-effector rot compensation limit [rad]
  - eefm_use_force_difference_control : Use force difference control or each limb force control. True by default.
  - eefm_swing_damping_force_thre : Swing damping control force threshold [N]
  - eefm_swing_damping_moment_thre : Swing damping control moment threshold [Nm]
  - use_zmp_truncation : Whether truncate zmp according to foot support polygon or not
- 足裏領域関連
  - eefm_leg_inside_margin : Inside foot margine [m]. Distance between foot end effector position and foot inside edge.
  - eefm_leg_outside_margin : Outside foot margine [m]. Distance between foot end effector position and foot outside edge.
  - eefm_leg_front_margin : Front foot margine [m]. Distance between foot end effector position and foot front edge
  - eefm_leg_rear_margin : Rear foot margine [m]. Distance between foot end effector position and foot rear edge.
  - eefm_support_polygon_vertices_sequence : Sequence of vertices for all end effectors assuming that the order is (rleg, lleg, rarm, larm)
- SwingEEModification関連
  - eefm_swing_rot_spring_gain : Sequence of all swing leg end-effector rotation spring gain (r,p,y).
  - eefm_swing_rot_time_const : Sequence of all swing leg end-effector rotation spring time constant [s] (r,p,y).
  - eefm_swing_pos_spring_gain : Sequence of all swing leg end-effector position spring gain (x,y,z).
  - eefm_swing_pos_time_const : Sequence of all swing leg end-effector position spring time constant [s] (x,y,z).
- BodyAttitudeControl関連
  - eefm_body_attitude_control_gain : Body attitude control gain [rad/s] (roll, pitch) for EEFM.
  - eefm_body_attitude_control_time_const : Time constant for body attitude control [s] (roll, pitch) for EEFM.
  - root_rot_compensation_limit : imit of compensation for difference between ref-act root rot [rad].
- ST関連
  - use_footguided_stabilizer
  - footguided_balance_time_const;
  - is_zmp_calc_enable : Bool sequence for all end effectors whether the end effector is used for zmp calculation . trueだと接触力からactZMPを求め、on_ground判定に利用
  - eefm_k1, eefm_k2, eefm_k3 : (use_act_state = false)
- IK関連
  - is_ik_enable : Bool sequence for all end effectors whether the end effector is used for limb IK
  - ik_limb_parameters
- Emergency関連
  - cp_check_margin : CP margin [m] (front, rear, inside, outside)
  - emergency_check_mode : CP or NO
- on_ground関連
  - contact_decision_threshold: on_ground判定のfz(world)の閾値. 5hzでフィルタして判定. (is_zmp_calc_enableのみ)
  - detection_time_to_air
- トルク制御関連
  - joint_control_mode
  - swing2landing_transition_time
  - landing_phase_time
  - landing2support_transition_time
  - support_phase_min_time
  - support2swing_transition_time
  - joint_servo_control_parameters;
- toe heel関連
  - toe/heel_pos/zmp_offset_x

**Comment by YutaKojio** [comment]
ほとんどいじらないものもあるけど，必要なのは網羅していると思います．

> leg_margin : Leg margin between foot end effector position and foot edge (front, rear, outside, inside) [m]. stride_limitation時の干渉判定と、CP転倒判定の閾値として用いる

leg_marginは干渉判定と，軌道生成時のZMPの支持領域への制限で使っています．
転倒判定で用いるのは，
ST paramの```eefm_support_polygon_vertices_sequence```および```cp_check_margin``` です．
